engine/src/mac-av-player.mm

```diff
@@ -862,6 +896,176 @@ - (void)setPlayer:(AVPlayer *)player
+{
+	_audioprocessingcontext *t_context;
+	t_context = (_audioprocessingcontext*) MTAudioProcessingTapGetStorage(tap);
+	if (processingFormat->mChannelsPerFrame == 2)
+	{
+		t_context->buffer_length = maxFrames * processingFormat->mBitsPerChannel / 8;
+		MCMemoryAllocate(4 * t_context->buffer_length, t_context->buffer);
+	}
+}
+
+void _audioprocessingtap_unprepare(MTAudioProcessingTapRef tap)
+{
+	_audioprocessingcontext *t_context;
+	t_context = (_audioprocessingcontext*) MTAudioProcessingTapGetStorage(tap);
+	if (t_context->buffer != nil)
+		MCMemoryDeallocate(t_context->buffer);
+	t_context->buffer = nil;
+}
+
+void _audioprocessingtap_process(MTAudioProcessingTapRef tap, CMItemCount numberFrames,
+	MTAudioProcessingTapFlags flags, AudioBufferList *bufferListInOut,
+	CMItemCount *numberFramesOut, MTAudioProcessingTapFlags *flagsOut)
+{
+	MTAudioProcessingTapGetSourceAudio(tap, numberFrames, bufferListInOut, flagsOut, nil, numberFramesOut);
+	
+	if (bufferListInOut->mNumberBuffers == 2)
+	{
+		_audioprocessingcontext *t_context;
+		t_context = (_audioprocessingcontext*)MTAudioProcessingTapGetStorage(tap);
+		
+		void *t_ll = t_context->buffer;
+		void *t_lr = ((uint8_t*)t_context->buffer) + t_context->buffer_length;
+		void *t_rl = ((uint8_t*)t_context->buffer) + t_context->buffer_length * 2;
+		void *t_rr = ((uint8_t*)t_context->buffer) + t_context->buffer_length * 3;
+
+		vDSP_vsmul((float*)bufferListInOut->mBuffers[0].mData, 1, &t_context->panning_filter->left_balance, (float*)t_ll, 1, *numberFramesOut);
+		vDSP_vsmul((float*)bufferListInOut->mBuffers[1].mData, 1, &t_context->panning_filter->right_balance, (float*)t_rr, 1, *numberFramesOut);
+
+		vDSP_vsmul((float*)bufferListInOut->mBuffers[0].mData, 1, &t_context->panning_filter->left_pan, (float*)t_rl, 1, *numberFramesOut);
+		vDSP_vsmul((float*)bufferListInOut->mBuffers[1].mData, 1, &t_context->panning_filter->right_pan, (float*)t_lr, 1, *numberFramesOut);
+
+		vDSP_vadd((float*)t_ll, 1, (float*)t_lr, 1, (float*)bufferListInOut->mBuffers[0].mData, 1, *numberFramesOut);
+		vDSP_vadd((float*)t_rl, 1, (float*)t_rr, 1, (float*)bufferListInOut->mBuffers[1].mData, 1, *numberFramesOut);
+	}
+}
+
+bool MCAVPlayerCreateAudioProcessingTap(MCAVPlayerPanningFilter *p_panning_filter, MTAudioProcessingTapRef &r_tap)
+{
+	MTAudioProcessingTapCallbacks t_callbacks;
+	t_callbacks.version = kMTAudioProcessingTapCallbacksVersion_0;
+	t_callbacks.clientInfo = p_panning_filter;
+	t_callbacks.init = _audioprocessingtap_init;
+	t_callbacks.prepare = _audioprocessingtap_prepare;
+	t_callbacks.process = _audioprocessingtap_process;
+	t_callbacks.unprepare = _audioprocessingtap_unprepare;
+	t_callbacks.finalize = _audioprocessingtap_finalize;
+
+	MTAudioProcessingTapRef t_tap;
+	OSStatus t_error;
+	t_error = MTAudioProcessingTapCreate(kCFAllocatorDefault, &t_callbacks, kMTAudioProcessingTapCreationFlag_PostEffects, &t_tap);
```

Comment:
@livecodeian we should check t_error here and return false on an error then check it below. At the moment we could be setting r_tap to the uninitialised t_tap here and calling CFRelease on it